src/jdk.naming.dns/share/classes/com/sun/jndi/dns/DnsClient.java

```diff
@@ -158,6 +159,16 @@ public DnsClient(String[] servers, int timeout, int retries)
             ne.setRootCause(e);
             throw ne;
         }
+        // register a cleaning action to close the Selector
+        // when this DNS client becomes phantom reachable
+        Selector sel = udpChannelSelector;
+        CleanerFactory.cleaner().register(this, () -> {
```

Comment:
I believe this is not quite right. The Cleanable returned by the `register` method should be saved in a final field. And close() below should call `Cleanable::clean` instead of `udpChannelSelector.close()`

Reply:
> I believe this is not quite right. The Cleanable returned by the `register` method should be saved in a final field. And close() below should call `Cleanable::clean` instead of `udpChannelSelector.close()`

Selector::close is idempotent so it doesn't really matter but your approach would mean the cleanable is deregistered when the API is used correctly, which is good thing.

Reply:
Thanks for the comments, implemented suggestions in 3a139b272b25d52b8be9593cee3aec3f373c12ff - new final field is added to save `Cleaner.Cleanable` instance for later use in `DnsClient::close` to close the selector and deregister the cleanable action.